src/vector_of_array.jl

```diff
@@ -719,7 +718,7 @@ end
 # for VectorOfArray with multi-dimensional parent arrays of arrays where all elements are the same type
 function Base.similar(vec::VectorOfArray{
         T, N, AT}) where {T, N, AT <: AbstractArray{<:AbstractArray{T}}}
-    return VectorOfArray(similar(Base.parent(vec)))
+    return VectorOfArray(similar.(Base.parent(vec)))
```

Comment:
Broadcasting `similar` here avoids some `#undef` issues. 

Side note: this implementation could replace these 3 functions https://github.com/SciML/RecursiveArrayTools.jl/blob/eacbe3fa275440ea863db85e63d627dd103205d6/src/vector_of_array.jl#L715-L729

However, doing so breaks these tests because `similar` can't be called on `<:Real` values: https://github.com/SciML/RecursiveArrayTools.jl/blob/eacbe3fa275440ea863db85e63d627dd103205d6/test/basic_indexing.jl#L221-L229
These tests don't seem correct to me; the parent array `x` isn't a recursive array (`eltype(x)` is not `<:AbstractArray`), so to me it doesn't make sense to test this broadcast edge case. Maybe I'm missing something though.

Reply:
Side note: this implementation could replace these 3 functions https://github.com/SciML/RecursiveArrayTools.jl/blob/eacbe3fa275440ea863db85e63d627dd103205d6/src/vector_of_array.jl#L715-L729

However, doing so breaks these tests because `similar` can't be called on `<:Real` values: https://github.com/SciML/RecursiveArrayTools.jl/blob/eacbe3fa275440ea863db85e63d627dd103205d6/test/basic_indexing.jl#L221-L229
These tests don't seem correct to me; the parent array `x` doesn't have underlying data structure `Vector{AbstractArray{T}}`, so it shouldn't be a valid edge case. Maybe I'm missing something?

Reply:
Numbers are also allowed. That's used a lot downstream.

Reply:
We can special-case that case where the elements are numbers, but we cannot throw it out.